src/commands/edge_app.rs

```diff
@@ -1,227 +1,777 @@
+           "description": manifest.description,
+           "icon": manifest.icon,
+           "author": manifest.author,
+           "homepage_url": manifest.homepage_url,
+           "file_tree": file_tree,
+        });
+
+        let response = commands::post(
+            &self.authentication,
+            "v4/edge-apps/versions?select=revision",
+            &json,
+        )?;
+
+        if let Some(arr) = response.as_array() {
+            if let Some(obj) = arr.get(0) {
+                if let Some(revision) = obj["revision"].as_u64() {
+                    debug!("New version revision: {}", revision);
+                    return Ok(revision as u32);
+                }
+            }
+        }
+
+        Err(CommandError::MissingField)
+    }
+
+    fn get_file_tree(
+        &self,
+        manifest: &EdgeAppManifest,
+    ) -> Result<HashMap<String, String>, CommandError> {
+        let response = commands::get(
+            &self.authentication,
+            &format!(
+                "v4/edge-apps/versions?select=file_tree&app_id=eq.{}&revision=eq.{}",
+                manifest.app_id, manifest.revision
+            ),
+        )?;
+
+        #[derive(Clone, Debug, Default, PartialEq, Deserialize)]
+        struct FileTree {
+            file_tree: HashMap<String, String>,
         }
 
+        let file_tree = serde_json::from_value::<Vec<FileTree>>(response)?;
+        if file_tree.is_empty() {
+            return Ok(HashMap::new());
+        }
+
+        Ok(file_tree[0].file_tree.clone())
+    }
+
+    fn requires_upload(
+        &self,
+        changed_settings: &SettingChanges,
+        changed_files: &FileChanges,
+    ) -> bool {
+        changed_settings.has_changes() || changed_files.has_changes()
+    }
+
+    fn get_version_asset_signatures(
+        &self,
+        manifest: &EdgeAppManifest,
+    ) -> Result<Vec<AssetSignature>, CommandError> {
+        Ok(serde_json::from_value(commands::get(
+            &self.authentication,
+            &format!(
+                "v4/assets?select=signature&app_id=eq.{}&app_revision=eq.{}&type=eq.edge-app-file",
+                manifest.app_id, manifest.revision
+            ),
+        )?)?)
+    }
+
+    fn ensure_assets_processing_finished(
+        &self,
+        manifest: &EdgeAppManifest,
+    ) -> Result<(), CommandError> {
+        const SLEEP_TIME: u64 = 2;
+        loop {
+            let value = commands::get(
+                &self.authentication,
+                &format!(
+                    "v4/assets?select=status&app_id=eq.{}&app_revision=eq.{}&status=neq.finished&limit=1",
+                    manifest.app_id, manifest.revision
+                ),
+            )?;
+            debug!("ensure_assets_processing_finished: {:?}", &value);
+
+            if let Some(array) = value.as_array() {
+                if array.is_empty() {
+                    break;
+                }
+            }
+            thread::sleep(Duration::from_secs(SLEEP_TIME));
```

Comment:
Let not sleep forever, may be fail after 20-30 secs?

Reply:
done